### PR TITLE
fix: type guard for sendable text-based channels

### PR DESCRIPTION
**Pull request by vladfrangu**
**Please describe the changes this PR makes and why it should be merged:**

Adds in `BaseChannel#isSendable`, which is a mix of `isTextBased()` and checking if you can actually send messages in the channel. This way, we can have the correct types for interaction.channel, and you can assert whether you got the interaction from a Group DM Channel or not

**Status and versioning classification:**

- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

**Comment by imnaiyar** [comment]
IG this (and the JSDoc) can also be updated to include PartialGroupDMChannel since it's considered text-based now
https://github.com/discordjs/discord.js/blob/4594896b5404c6a34e07544951c59ff8f3657184/packages/discord.js/src/util/Constants.js#L122

File: packages/discord.js/src/structures/BaseChannel.js [hunk 1]
```diff
@@ -155,6 +155,14 @@ class BaseChannel extends Base {
     return 'availableTags' in this;
   }
 
+  /**
+   * Indicates whether this channel is {@link TextBasedChannels text-based} and can have messages sent in it.
+   * @returns {boolean}
+   */
+  isSendable() {
+    return this.isTextBased() && 'send' in this;
+  }
```
Comment (Qjuh) [inline packages/discord.js/src/structures/BaseChannel.js:164]: Is the first check here needed? If a channel had a send method it should be considered sendable. Even if for whatever reason it‘s not TextBased (although no channel type that does that currently exists).

File: packages/discord.js/typings/index.d.ts [hunk 2]
```diff
@@ -6878,11 +6880,15 @@ export type Channel =
 
 export type TextBasedChannel = Exclude<Extract<Channel, { type: TextChannelType }>, ForumChannel | MediaChannel>;
 
+export type SendableTextBasedChannels = Exclude<TextBasedChannel, PartialGroupDMChannel>;
```
Comment (Qjuh) [inline packages/discord.js/typings/index.d.ts:6883]: If TextBased check is dropped then this would simply be `SendableChannels`.

**Review by Qjuh — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/structures/BaseChannel.js (hunk 1 above)
Comment (vladfrangu) [reply packages/discord.js/src/structures/BaseChannel.js:164]: I'd rather keep it in, if in the future for god-knows what reason we get a channel with a .send that doesn't do messages XD

Comment (Qjuh) [reply packages/discord.js/src/structures/BaseChannel.js:164]: Then you can simply chain both typeguards as a user… but I‘m good with either way

File: packages/discord.js/typings/index.test-d.ts [hunk 3]
```diff
@@ -2593,3 +2594,16 @@ declare const poll: Poll;
...
+  if (!interaction.channel) {
+    return;
+  }
+
+  // @ts-expect-error
+  interaction.channel.send();
+
+  if (interaction.channel.isSendable()) {
+    expectType<SendableTextBasedChannels>(interaction.channel);
+  }
```
Comment (Qjuh) [inline packages/discord.js/typings/index.test-d.ts:2608]: To not only check typeguard with type but also actual purpose, maybe add: 
```suggestion
    interaction.channel.send()
  }
```

**Review by Qjuh — APPROVED** [review]

File: packages/discord.js/src/structures/BaseChannel.js (hunk 1 above)
Comment (Jiralite) [reply packages/discord.js/src/structures/BaseChannel.js:164]: Leaning towards @Qjuh, but I am majorly impartial here.

**Review by Jiralite — APPROVED** [review]

**Review by didinele — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

**Review by Qjuh — APPROVED** [review]

**Pull request merged**